File: qt/gen_qsystemsemaphore.go
package qt

import (
	"unsafe"
)

type QSystemSemaphore__AccessMode int

const (
	QSystemSemaphore__Open   QSystemSemaphore__AccessMode = 0
	QSystemSemaphore__Create QSystemSemaphore__AccessMode = 1
)

type QSystemSemaphore__SystemSemaphoreError int

const (
	QSystemSemaphore__NoError          QSystemSemaphore__SystemSemaphoreError = 0
	QSystemSemaphore__PermissionDenied QSystemSemaphore__SystemSemaphoreError = 1
	QSystemSemaphore__KeyError         QSystemSemaphore__SystemSemaphoreError = 2
	QSystemSemaphore__AlreadyExists    QSystemSemaphore__SystemSemaphoreError = 3
	QSystemSemaphore__NotFound         QSystemSemaphore__SystemSemaphoreError = 4
	QSystemSemaphore__OutOfResources   QSystemSemaphore__SystemSemaphoreError = 5
	QSystemSemaphore__UnknownError     QSystemSemaphore__SystemSemaphoreError = 6
)

type QSystemSemaphore struct {
	h          uintptr
	isSubclass bool
}

// NewQSystemSemaphore constructs a new QSystemSemaphore object.
func NewQSystemSemaphore(key *QNativeIpcKey) *QSystemSemaphore {
	ret := newQSystemSemaphore(QSystemSemaphore_new(key.cPointer()))
	ret.isSubclass = true
	return ret
}

// NewQSystemSemaphore2 constructs a new QSystemSemaphore object.
func NewQSystemSemaphore2(key string) *QSystemSemaphore {
	key_ms := struct_miqt_string{}
	key_ms.data = CString(key)
	key_ms.len = size_t(len(key))
	defer free(unsafe.Pointer(key_ms.data))

	ret := newQSystemSemaphore(QSystemSemaphore_new2(key_ms))
	ret.isSubclass = true
	return ret
}

// NewQSystemSemaphore3 constructs a new QSystemSemaphore object.
func NewQSystemSemaphore3(key *QNativeIpcKey, initialValue int) *QSystemSemaphore {
	ret := newQSystemSemaphore(QSystemSemaphore_new3(key.cPointer(), (int)(initialValue)))
	ret.isSubclass = true
	return ret
}

// NewQSystemSemaphore4 constructs a new QSystemSemaphore object.
func NewQSystemSemaphore4(key *QNativeIpcKey, initialValue int, param3 AccessMode) *QSystemSemaphore {
	ret := newQSystemSemaphore(QSystemSemaphore_new4(key.cPointer(), (int)(initialValue), param3))
	ret.isSubclass = true
	return ret
}

// NewQSystemSemaphore5 constructs a new QSystemSemaphore object.
func NewQSystemSemaphore5(key string, initialValue int) *QSystemSemaphore {
	key_ms := struct_miqt_string{}
	key_ms.data = CString(key)
	key_ms.len = size_t(len(key))
	defer free(unsafe.Pointer(key_ms.data))

	ret := newQSystemSemaphore(QSystemSemaphore_new5(key_ms, (int)(initialValue)))
	ret.isSubclass = true
	return ret
}

// NewQSystemSemaphore6 constructs a new QSystemSemaphore object.
func NewQSystemSemaphore6(key string, initialValue int, mode AccessMode) *QSystemSemaphore {
	key_ms := struct_miqt_string{}
	key_ms.data = CString(key)
	key_ms.len = size_t(len(key))
	defer free(unsafe.Pointer(key_ms.data))

	ret := newQSystemSemaphore(QSystemSemaphore_new6(key_ms, (int)(initialValue), mode))
	ret.isSubclass = true
	return ret
}

func QSystemSemaphore_Tr(sourceText string) string {
	sourceText_Cstring := CString(sourceText)
	defer free(unsafe.Pointer(sourceText_Cstring))
	var _ms struct_miqt_string = QSystemSemaphore_Tr(sourceText_Cstring)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QSystemSemaphore) SetNativeKey(key *QNativeIpcKey) {
	QSystemSemaphore_SetNativeKey(this.h, key.cPointer())
}

func (this *QSystemSemaphore) SetNativeKeyWithKey(key string) {
	key_ms := struct_miqt_string{}
	key_ms.data = CString(key)
	key_ms.len = size_t(len(key))
	defer free(unsafe.Pointer(key_ms.data))
	QSystemSemaphore_SetNativeKeyWithKey(this.h, key_ms)
}

func (this *QSystemSemaphore) NativeIpcKey() *QNativeIpcKey {
	_goptr := newQNativeIpcKey(QSystemSemaphore_NativeIpcKey(this.h))
	_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
	return _goptr
}

func (this *QSystemSemaphore) SetKey(key string) {
	key_ms := struct_miqt_string{}
	key_ms.data = CString(key)
	key_ms.len = size_t(len(key))
	defer free(unsafe.Pointer(key_ms.data))
	QSystemSemaphore_SetKey(this.h, key_ms)
}

func (this *QSystemSemaphore) Key() string {
	var _ms struct_miqt_string = QSystemSemaphore_Key(this.h)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QSystemSemaphore) Acquire() bool {
	return (bool)(QSystemSemaphore_Acquire(this.h))
}

func (this *QSystemSemaphore) Release() bool {
	return (bool)(QSystemSemaphore_Release(this.h))
}

func (this *QSystemSemaphore) Error() SystemSemaphoreError {
	xxxxxxxxx
}

func (this *QSystemSemaphore) ErrorString() string {
	var _ms struct_miqt_string = QSystemSemaphore_ErrorString(this.h)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func QSystemSemaphore_IsKeyTypeSupported(typeVal QNativeIpcKey__Type) bool {
	return (bool)(QSystemSemaphore_IsKeyTypeSupported((uint16_t)(typeVal)))
}

func QSystemSemaphore_PlatformSafeKey(key string) *QNativeIpcKey {
	key_ms := struct_miqt_string{}
	key_ms.data = CString(key)
	key_ms.len = size_t(len(key))
	defer free(unsafe.Pointer(key_ms.data))
	_goptr := newQNativeIpcKey(QSystemSemaphore_PlatformSafeKey(key_ms))
	_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
	return _goptr
}

func QSystemSemaphore_LegacyNativeKey(key string) *QNativeIpcKey {
	key_ms := struct_miqt_string{}
	key_ms.data = CString(key)
	key_ms.len = size_t(len(key))
	defer free(unsafe.Pointer(key_ms.data))
	_goptr := newQNativeIpcKey(QSystemSemaphore_LegacyNativeKey(key_ms))
	_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
	return _goptr
}

func QSystemSemaphore_Tr2(sourceText string, disambiguation string) string {
	sourceText_Cstring := CString(sourceText)
	defer free(unsafe.Pointer(sourceText_Cstring))
	disambiguation_Cstring := CString(disambiguation)
	defer free(unsafe.Pointer(disambiguation_Cstring))
	var _ms struct_miqt_string = QSystemSemaphore_Tr2(sourceText_Cstring, disambiguation_Cstring)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func QSystemSemaphore_Tr3(sourceText string, disambiguation string, n int) string {
	sourceText_Cstring := CString(sourceText)
	defer free(unsafe.Pointer(sourceText_Cstring))
	disambiguation_Cstring := CString(disambiguation)
	defer free(unsafe.Pointer(disambiguation_Cstring))
	var _ms struct_miqt_string = QSystemSemaphore_Tr3(sourceText_Cstring, disambiguation_Cstring, (int)(n))
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QSystemSemaphore) SetNativeKey2(key *QNativeIpcKey, initialValue int) {
	QSystemSemaphore_SetNativeKey2(this.h, key.cPointer(), (int)(initialValue))
}

func (this *QSystemSemaphore) SetNativeKey3(key *QNativeIpcKey, initialValue int, param3 AccessMode) {
	QSystemSemaphore_SetNativeKey3(this.h, key.cPointer(), (int)(initialValue), param3)
}

func (this *QSystemSemaphore) SetNativeKey22(key string, initialValue int) {
	key_ms := struct_miqt_string{}
	key_ms.data = CString(key)
	key_ms.len = size_t(len(key))
	defer free(unsafe.Pointer(key_ms.data))
	QSystemSemaphore_SetNativeKey22(this.h, key_ms, (int)(initialValue))
}

func (this *QSystemSemaphore) SetNativeKey32(key string, initialValue int, mode AccessMode) {
	key_ms := struct_miqt_string{}
	key_ms.data = CString(key)
	key_ms.len = size_t(len(key))
	defer free(unsafe.Pointer(key_ms.data))
	QSystemSemaphore_SetNativeKey32(this.h, key_ms, (int)(initialValue), mode)
}

func (this *QSystemSemaphore) SetNativeKey4(key string, initialValue int, mode AccessMode, typeVal QNativeIpcKey__Type) {
	key_ms := struct_miqt_string{}
	key_ms.data = CString(key)
	key_ms.len = size_t(len(key))
	defer free(unsafe.Pointer(key_ms.data))
	QSystemSemaphore_SetNativeKey4(this.h, key_ms, (int)(initialValue), mode, (uint16_t)(typeVal))
}

func (this *QSystemSemaphore) SetKey2(key string, initialValue int) {
	key_ms := struct_miqt_string{}
	key_ms.data = CString(key)
	key_ms.len = size_t(len(key))
	defer free(unsafe.Pointer(key_ms.data))
	QSystemSemaphore_SetKey2(this.h, key_ms, (int)(initialValue))
}

func (this *QSystemSemaphore) SetKey3(key string, initialValue int, mode AccessMode) {
	key_ms := struct_miqt_string{}
	key_ms.data = CString(key)
	key_ms.len = size_t(len(key))
	defer free(unsafe.Pointer(key_ms.data))
	QSystemSemaphore_SetKey3(this.h, key_ms, (int)(initialValue), mode)
}

func (this *QSystemSemaphore) Release1(n int) bool {
	return (bool)(QSystemSemaphore_Release1(this.h, (int)(n)))
}

func QSystemSemaphore_PlatformSafeKey2(key string, typeVal QNativeIpcKey__Type) *QNativeIpcKey {
	key_ms := struct_miqt_string{}
	key_ms.data = CString(key)
	key_ms.len = size_t(len(key))
	defer free(unsafe.Pointer(key_ms.data))
	_goptr := newQNativeIpcKey(QSystemSemaphore_PlatformSafeKey2(key_ms, (uint16_t)(typeVal)))
	_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
	return _goptr
}

func QSystemSemaphore_LegacyNativeKey2(key string, typeVal QNativeIpcKey__Type) *QNativeIpcKey {
	key_ms := struct_miqt_string{}
	key_ms.data = CString(key)
	key_ms.len = size_t(len(key))
	defer free(unsafe.Pointer(key_ms.data))
	_goptr := newQNativeIpcKey(QSystemSemaphore_LegacyNativeKey2(key_ms, (uint16_t)(typeVal)))
	_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
	return _goptr
}

File: qt/gen_qsystemsemaphore.cpp
// +build ignore

#include <QNativeIpcKey>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <QSystemSemaphore>
#include <qsystemsemaphore.h>
#include "gen_qsystemsemaphore.h"

QSystemSemaphore* QSystemSemaphore_new(QNativeIpcKey* key) {
	return new QSystemSemaphore(*key);
}

QSystemSemaphore* QSystemSemaphore_new2(struct miqt_string key) {
	QString key_QString = QString::fromUtf8(key.data, key.len);
	return new QSystemSemaphore(key_QString);
}

QSystemSemaphore* QSystemSemaphore_new3(QNativeIpcKey* key, int initialValue) {
	return new QSystemSemaphore(*key, static_cast<int>(initialValue));
}

QSystemSemaphore* QSystemSemaphore_new4(QNativeIpcKey* key, int initialValue, AccessMode param3) {
	return new QSystemSemaphore(*key, static_cast<int>(initialValue), param3);
}

QSystemSemaphore* QSystemSemaphore_new5(struct miqt_string key, int initialValue) {
	QString key_QString = QString::fromUtf8(key.data, key.len);
	return new QSystemSemaphore(key_QString, static_cast<int>(initialValue));
}

QSystemSemaphore* QSystemSemaphore_new6(struct miqt_string key, int initialValue, AccessMode mode) {
	QString key_QString = QString::fromUtf8(key.data, key.len);
	return new QSystemSemaphore(key_QString, static_cast<int>(initialValue), mode);
}

struct miqt_string QSystemSemaphore_Tr(const char* sourceText) {
	QString _ret = QSystemSemaphore::tr(sourceText);
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}

void QSystemSemaphore_SetNativeKey(QSystemSemaphore* self, QNativeIpcKey* key) {
	self->setNativeKey(*key);
}

void QSystemSemaphore_SetNativeKeyWithKey(QSystemSemaphore* self, struct miqt_string key) {
	QString key_QString = QString::fromUtf8(key.data, key.len);
	self->setNativeKey(key_QString);
}

QNativeIpcKey* QSystemSemaphore_NativeIpcKey(const QSystemSemaphore* self) {
	return new QNativeIpcKey(self->nativeIpcKey());
}

void QSystemSemaphore_SetKey(QSystemSemaphore* self, struct miqt_string key) {
	QString key_QString = QString::fromUtf8(key.data, key.len);
	self->setKey(key_QString);
}

struct miqt_string QSystemSemaphore_Key(const QSystemSemaphore* self) {
	QString _ret = self->key();
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}

bool QSystemSemaphore_Acquire(QSystemSemaphore* self) {
	return self->acquire();
}

bool QSystemSemaphore_Release(QSystemSemaphore* self) {
	return self->release();
}

SystemSemaphoreError QSystemSemaphore_Error(const QSystemSemaphore* self) {
	return self->error();
}

struct miqt_string QSystemSemaphore_ErrorString(const QSystemSemaphore* self) {
	QString _ret = self->errorString();
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}

bool QSystemSemaphore_IsKeyTypeSupported(uint16_t typeVal) {
	return QSystemSemaphore::isKeyTypeSupported(static_cast<QNativeIpcKey::Type>(typeVal));
}

QNativeIpcKey* QSystemSemaphore_PlatformSafeKey(struct miqt_string key) {
	QString key_QString = QString::fromUtf8(key.data, key.len);
	return new QNativeIpcKey(QSystemSemaphore::platformSafeKey(key_QString));
}

QNativeIpcKey* QSystemSemaphore_LegacyNativeKey(struct miqt_string key) {
	QString key_QString = QString::fromUtf8(key.data, key.len);
	return new QNativeIpcKey(QSystemSemaphore::legacyNativeKey(key_QString));
}

struct miqt_string QSystemSemaphore_Tr2(const char* sourceText, const char* disambiguation) {
	QString _ret = QSystemSemaphore::tr(sourceText, disambiguation);
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}

struct miqt_string QSystemSemaphore_Tr3(const char* sourceText, const char* disambiguation, int n) {
	QString _ret = QSystemSemaphore::tr(sourceText, disambiguation, static_cast<int>(n));
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}

void QSystemSemaphore_SetNativeKey2(QSystemSemaphore* self, QNativeIpcKey* key, int initialValue) {
	self->setNativeKey(*key, static_cast<int>(initialValue));
}

void QSystemSemaphore_SetNativeKey3(QSystemSemaphore* self, QNativeIpcKey* key, int initialValue, AccessMode param3) {
	self->setNativeKey(*key, static_cast<int>(initialValue), param3);
}

void QSystemSemaphore_SetNativeKey22(QSystemSemaphore* self, struct miqt_string key, int initialValue) {
	QString key_QString = QString::fromUtf8(key.data, key.len);
	self->setNativeKey(key_QString, static_cast<int>(initialValue));
}

void QSystemSemaphore_SetNativeKey32(QSystemSemaphore* self, struct miqt_string key, int initialValue, AccessMode mode) {
	QString key_QString = QString::fromUtf8(key.data, key.len);
	self->setNativeKey(key_QString, static_cast<int>(initialValue), mode);
}

void QSystemSemaphore_SetNativeKey4(QSystemSemaphore* self, struct miqt_string key, int initialValue, AccessMode mode, uint16_t typeVal) {
	QString key_QString = QString::fromUtf8(key.data, key.len);
	self->setNativeKey(key_QString, static_cast<int>(initialValue), mode, static_cast<QNativeIpcKey::Type>(typeVal));
}

void QSystemSemaphore_SetKey2(QSystemSemaphore* self, struct miqt_string key, int initialValue) {
	QString key_QString = QString::fromUtf8(key.data, key.len);
	self->setKey(key_QString, static_cast<int>(initialValue));
}

void QSystemSemaphore_SetKey3(QSystemSemaphore* self, struct miqt_string key, int initialValue, AccessMode mode) {
	QString key_QString = QString::fromUtf8(key.data, key.len);
	self->setKey(key_QString, static_cast<int>(initialValue), mode);
}

bool QSystemSemaphore_Release1(QSystemSemaphore* self, int n) {
	return self->release(static_cast<int>(n));
}

QNativeIpcKey* QSystemSemaphore_PlatformSafeKey2(struct miqt_string key, uint16_t typeVal) {
	QString key_QString = QString::fromUtf8(key.data, key.len);
	return new QNativeIpcKey(QSystemSemaphore::platformSafeKey(key_QString, static_cast<QNativeIpcKey::Type>(typeVal)));
}

QNativeIpcKey* QSystemSemaphore_LegacyNativeKey2(struct miqt_string key, uint16_t typeVal) {
	QString key_QString = QString::fromUtf8(key.data, key.len);
	return new QNativeIpcKey(QSystemSemaphore::legacyNativeKey(key_QString, static_cast<QNativeIpcKey::Type>(typeVal)));
}

void QSystemSemaphore_Delete(QSystemSemaphore* self, bool isSubclass) {
	if (isSubclass) {
		delete dynamic_cast<QSystemSemaphore*>( self );
	} else {
		delete self;
	}
}

